### PR TITLE
feat(SCT-263): add new use case for adding a personal relationship

### PR DESCRIPTION
**Pull request by chubberlisk**
## Link to JIRA ticket

https://hackney.atlassian.net/browse/SCT-263

### *What is the problem we're trying to solve*

We want to be able to add a personal relationship. In #345, we added methods in the `DatabaseGateway` to support this new use case.

### *What changes have we introduced*

This PR adds a new use case: `PersonalRelationship#ExecutePost` that will ensure that:

- the provided `personId` does exist
- the provided `otherPersonId` does exist
- the provided `description` of type does exist
- the provided relationship doesn't already exist for the same person with the same type

It will then create a new personal relationship for the provided person and create the inverse relationship.

#### _Checklist_

- [ ] Added end-to-end (i.e. integration) tests where necessary e.g to test complete functionality of a newly added feature
- [x] Added tests to cover all new production code
- [x] Added comments to the README or updated relevant documentation _(add link to documentation)_, where necessary.
- [x] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

### *Follow up actions after merging PR*

Create the API endpoint: `api/v1/relationships/personal`.

### CODE REVIEW

File: SocialCareCaseViewerApi/V1/UseCase/PersonalRelationshipsUseCase.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+using SocialCareCaseViewerApi.V1.Boundary.Requests;
+using SocialCareCaseViewerApi.V1.Exceptions;
+using SocialCareCaseViewerApi.V1.Gateways;
+using SocialCareCaseViewerApi.V1.UseCase.Interfaces;
+
+namespace SocialCareCaseViewerApi.V1.UseCase
+{
+    public class PersonalRelationshipsUseCase : IPersonalRelationshipsUseCase
+    {
+        private IDatabaseGateway _databaseGateway;
```
Comment (JohnFarrellDev) [inline SocialCareCaseViewerApi/V1/UseCase/PersonalRelationshipsUseCase.cs:11]: can be `readonly`

File: SocialCareCaseViewerApi/V1/UseCase/PersonalRelationshipsUseCase.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+            var typeDoesNotExist = type == null;
+            if (typeDoesNotExist) throw new PersonalRelationshipTypeNotFoundException($"\"type\" with \"{request.Type}\" was not found.");
+
+            var personWithPersonalRelationships = _databaseGateway.GetPersonWithPersonalRelationshipsByPersonId(request.PersonId);
+
+            var personalRelationships = personWithPersonalRelationships.PersonalRelationships;
+            var personalRelationshipAlreadyExists = personalRelationships.Find(pr => pr.OtherPersonId == request.OtherPersonId && pr.Type.Description == request.Type) != null;
+            if (personalRelationshipAlreadyExists) throw new PersonalRelationshipAlreadyExistsException($"Personal relationship with \"type\" of \"{request.Type}\" already exists.");
+
+            _databaseGateway.CreatePersonalRelationship(request);
```
Comment (JohnFarrellDev) [inline SocialCareCaseViewerApi/V1/UseCase/PersonalRelationshipsUseCase.cs:39]: If the request.TypeId is null the CreatePersonalRelationship function will throw an exception I think when it tries to cast null value to type long.

File: SocialCareCaseViewerApi/V1/UseCase/PersonalRelationshipsUseCase.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+            _databaseGateway.CreatePersonalRelationship(new CreatePersonalRelationshipRequest()
+            {
+                PersonId = request.OtherPersonId,
+                OtherPersonId = request.PersonId,
+                TypeId = type.InverseTypeId,
+                IsMainCarer = null,
+                IsInformalCarer = null,
+                Details = null
```
Comment (JohnFarrellDev) [inline SocialCareCaseViewerApi/V1/UseCase/PersonalRelationshipsUseCase.cs:48]: The CreatePersonalRelationshipRequest object states these values can't be null.

You could make them nullable but keep the FluentValidation checks to ensure the actual API request validates they aren't null. This way you know the request actually has assigned the values but can also set the value as null directly yourself without compiler warnings.

File: SocialCareCaseViewerApi.Tests/V1/UseCase/PersonalRelationships/PersonalRelationshipsExecutePostUseCaseTests.cs [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+            _personalRelationshipsUseCase.ExecutePost(_request);
+
+            _mockDatabaseGateway.Verify(gateway => gateway.GetPersonalRelationshipTypeByDescription(_request.Type));
+        }
+
+        [Test]
+        public void WhenTypeDoesNotExistThrowsPersonalRelationshipTypeNotFoundException()
+        {
+            _mockDatabaseGateway.Setup(x => x.GetPersonalRelationshipTypeByDescription(It.IsAny<string>()))
+                .Returns((PersonalRelationshipType) null);
```
Comment (JohnFarrellDev) [inline SocialCareCaseViewerApi.Tests/V1/UseCase/PersonalRelationships/PersonalRelationshipsExecutePostUseCaseTests.cs:100]: If you want your mock to return null you can just leave off the `.Returns(...)`

File: SocialCareCaseViewerApi.Tests/V1/UseCase/PersonalRelationships/PersonalRelationshipsExecutePostUseCaseTests.cs [hunk 5]
```diff
@@ -0,0 +1,156 @@
+using FluentAssertions;
```
Comment (JohnFarrellDev) [inline SocialCareCaseViewerApi.Tests/V1/UseCase/PersonalRelationships/PersonalRelationshipsExecutePostUseCaseTests.cs:1]: Good very comprehensive tests :)

**Review by JohnFarrellDev — APPROVED** [review]
:rocket:

File: SocialCareCaseViewerApi.Tests/V1/UseCase/PersonalRelationships/PersonalRelationshipsExecutePostUseCaseTests.cs (hunk 5 above)
Comment (chubberlisk) [reply SocialCareCaseViewerApi.Tests/V1/UseCase/PersonalRelationships/PersonalRelationshipsExecutePostUseCaseTests.cs:1]: Not going to lie, I enjoyed TDDing this very much! 🕺

File: SocialCareCaseViewerApi.Tests/V1/UseCase/PersonalRelationships/PersonalRelationshipsExecutePostUseCaseTests.cs (hunk 4 above)
Comment (chubberlisk) [reply SocialCareCaseViewerApi.Tests/V1/UseCase/PersonalRelationships/PersonalRelationshipsExecutePostUseCaseTests.cs:100]: Ahhhh that's true! But I think I'd prefer to keep it and have the test be explicit about it for readability and clarity.

File: SocialCareCaseViewerApi/V1/UseCase/PersonalRelationshipsUseCase.cs (hunk 1 above)
Comment (chubberlisk) [reply SocialCareCaseViewerApi/V1/UseCase/PersonalRelationshipsUseCase.cs:11]: I need to get VS Code to pick these up! 🤦

File: SocialCareCaseViewerApi/V1/UseCase/PersonalRelationshipsUseCase.cs (hunk 2 above)
Comment (chubberlisk) [reply SocialCareCaseViewerApi/V1/UseCase/PersonalRelationshipsUseCase.cs:39]: m8, good spot! I completely forgot I needed to do this... I did it for the inverse but not for this one. 🤦

File: SocialCareCaseViewerApi/V1/UseCase/PersonalRelationshipsUseCase.cs (hunk 3 above)
Comment (chubberlisk) [reply SocialCareCaseViewerApi/V1/UseCase/PersonalRelationshipsUseCase.cs:50]: That makes sense, I think I keep going confused by `null!`. Could you check that the latest commit is correct please?

Comment (JohnFarrellDev) [reply SocialCareCaseViewerApi/V1/UseCase/PersonalRelationshipsUseCase.cs:50]: So if something is nullable (got the ? after the type) we don't need `null!`.

https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-forgiving

My understanding is saying `null!` is basically telling the compiler I know this isn't meant to be null but I know right now I'm not assigning this a value so for now it is null and that's ok.

Comment (chubberlisk) [reply SocialCareCaseViewerApi/V1/UseCase/PersonalRelationshipsUseCase.cs:50]: Ahhh okay, I got! Your explanation is much clearer than the docs for me. 🙇‍♀️

Will remove `null!`.

**Pull request merged**